doc/sonic-vrf-hld-v0.2.md

```diff
@@ -0,0 +1,343 @@
+
+1.  References
+
+- VRF feature Requirement
+
+    - Add or Delete VRF instance
+    - Bind VRF to a L3 interface.
+    - Static IP route with VRF
+    - Enable eBGP/OSPF VRF aware in SONiC
+    - Support fall through lookup
+    - TBD: VRF route leaking between VRFs.
+
+Note: linux kernel use VRF master device to support VRF and it supports admin
+up/down on VRF maser dev. But we don't plan to support it on SONIC.
+
+Dependencies
+============
+
+VRF feature needs the following software package/upgrade
+
+1.  Linux kernel 4.9
+
+Linux Kernel 4.9 support generic IP VRF with L3 master net device. Every L3
+master net device has its own FIB. The name of the master device is the
+VRF’s name. Real network interface can join the VRF by becoming the slave of
+the master net device.
+
+Application can get creation of deletion of VRF master device via RTNETLINK,
+as well as information about slave net device joining a VRF.
+
+Linux kernel supports VRF forwarding using PBR scheme. It will fall to main
+routing table to check IP lookup. VRF also can have its own default network
+instruction in case VRF lookup fails.
+
+2.  FRRouting is needed to support BGP/OSPF VRF aware routing.
+
+3.  IProute2 version should be ss161212 or later to support iproute2 CLIs to
+    configure the switch.
+
+Example of using iprout2:
+
+```
+VRF name: vrf-blue，fib-table-id: 10
+
+$ ip link add name vrf-blue type vrf table 10
+
+//enable VRF
+
+$ ip link set dev vrf-blue up
+
+// disable global VRF lookup
+
+$ ip [-6] route add table 10 unreachable default
+
+//binding sw1p3 device to vrf-blue
+
+$ ip link set dev sw1p3 master vrf-blue
+
+// descend local table pref
+
+ip [-6] rule add pref 32765 table local && ip [-6] rule del pref 0
+```
+
+4.  SAI VRF support
+
+SAI right now does not seem having VRF concept, it does have VR.
+
+We propose to implement VR as “virtual router” and VRF as “virtual router
+forwarding”
+
+VR is defined as an logical routing system. VRF is defined as forwarding
+domain within a VR.
+
+As this stage, we assume one VR per system. Only implement VRFs within this
+VR.
+
+Accordingly, we need to add vrf_id to sai_Route_entry and add vrf attribute
+to sai_routeInterface object.
+
+An alternative method is using VR as VRF. But it is needed to add two
+attribution to VR object.
+
+```
+/*
+ * \@brief if it is global vrf
+ *
+ * \@type bool
+ * \@flags CREATE_AND_SET
+ * \@default true
+ */
+ SAI_VIRTUAL_ROUTER_ATTR_GLOBAL
+
+/*
+ * \@brief continue to do global fib lookup while current vrf fib lookup
+ *  missed
+ * 
+ * \@type bool
+ * \@flags CREATE_AND_SET
+ * \@default false
+ */
+ SAI_VIRTUAL_ROUTER_ATTR_FALL_THROUGH
+```
+
+SONiC system diagram for VRF
+============================
+
+The following is high level diagram of modules with VRF support.
+
+## The schema changes
+
+1. Adding VRF related configuration in config_db.json
+
+```
+"VRF": {
+    "VRF-blue": {
+        "fall_through":"1" //enable global fib lookup while vrf fib lookup missed
```

Comment:
Please change fall_through to fallback. The term fall_through tends to be used for routemaps but for vrfs the industry established term is fallback.

There should be no restriction that the fallback be the global/main table even if that may be the only one we plan to support (and I don't see why we plan to support only global/main as fallback).

Your document needs to specify how the fallback is specified. I assume the table id will be used since linux doesn't have a name for the global/main table.

Reply:
@nikos-github Fallback term is better.I will revise it. The fallback global feature is very useful for specify VRF user access internet through global/main route which is defined by RFC4364. Some enterprise user still use this to access internet on vpn environment. Different vrf route leak is beyond this document scope.